rclcpp/src/rclcpp/executable_list.cpp

```diff
@@ -24,3 +24,57 @@ ExecutableList::ExecutableList()
 
 ExecutableList::~ExecutableList()
 {}
+
+void
+ExecutableList::clear()
+{
+  this->timer.clear();
+  this->number_of_timers = 0;
+
+  this->subscription.clear();
+  this->number_of_subscriptions = 0;
+
+  this->service.clear();
+  this->number_of_services = 0;
+
+  this->client.clear();
+  this->number_of_clients = 0;
+
+  this->waitable.clear();
+  this->number_of_waitables = 0;
+}
+
+void
+ExecutableList::add_subscription(rclcpp::SubscriptionBase::SharedPtr subscription)
+{
+  this->subscription.push_back(std::move(subscription));
```

Comment:
Why move here?

Reply:
As `subscription` is passed by value this `std::move` avoids the slow increasing/decreasing of the atomic reference count. We "steal" the `subscription` atomic reference count and we nullify its `shared_ptr`.

Reply:
But move semantics will cause this to happen anyways. My point was that the move is redundant, in my opinion. I suppose it is fine to leave it, but I do think it is redundant.

Reply:
See 6b15351

Reply:
https://stackoverflow.com/a/29644158/7108533

Reply:
@alsora I don't know what that answer is supposed to indicate. I know moving it is an optimization, but my point is that the compiler knows `subscription` is a local variable and not used after line 50, so it will move it automatically AFAIK. Due to copy elision https://en.cppreference.com/w/cpp/language/copy_elision, but maybe that's only guaranteed in C++17...

My understanding is that the situation is like this:

```c++
void foo(std::string pong);

void bar(std::string ping) {
  foo(ping);  // this automatically moves s into the input parameter of foo()
  // and further more is preferable to this:
  // foo(std::move(ping));
}

void main(void) {
  std::string something = "something";
  bar(something);  // copy occurs making 'ping', but 'ping' is moved into 'pong'
}
```

But maybe I'm wrong.

Reply:
As far as I know, in C++ 17 you are guaranteed that copy elision will happen in that situation, but in C++14 it may or may not happen.

Reply:
https://www.onlinegdb.com/SysgHtoLL did a quick test, using `C++17` and debugging the pointer.
Without `std::move`, the atomic counter is incremented/decremented.